File: cpp/Vector2f.cpp
//
// Created by Tamir Eliraz on 11/2/2024.
//

#include "Vector2f.h"

Geography::Vector2f::Vector2f() : Vector2f(0.0F) { }

Geography::Vector2f::Vector2f(float xy) : Vector2f(xy, xy) { }

Geography::Vector2f::Vector2f(float x, float y) : x(x), y(y) { }

Geography::Vector2f &Geography::Vector2f::operator+=(const Geography::Vector2f &other) {
    x += other.x;
    y += other.y;
    return *this;
}

Geography::Vector2f &Geography::Vector2f::operator-=(const Geography::Vector2f &other) {
    x -= other.x;
    y -= other.y;
    return *this;
}

Geography::Vector2f &Geography::Vector2f::operator*=(float c) {
    x *= c;
    y *= c;
    return *this;
}

Geography::Vector2f Geography::operator+(const Geography::Vector2f &lhs, const Geography::Vector2f &rhs) {
    return Vector2f(lhs.x + rhs.x, lhs.y + rhs.y);
}

Geography::Vector2f Geography::operator-(const Geography::Vector2f &lhs, const Geography::Vector2f &rhs) {
    return Vector2f(lhs.x - rhs.x, lhs.y - rhs.y);
}

Geography::Vector2f Geography::operator*(const Geography::Vector2f &lhs, float c) {
    return Vector2f(lhs.x - c, lhs.y - c);
}

std::ostream &Geography::operator<<(std::ostream &out, const Geography::Vector2f &rhs) {
    out << "(" << rhs.x << ", " << rhs.y << ")";
    return out;
}

float Geography::Vector2f::getX() const {
    return x;
}

Geography::Vector2f &Geography::Vector2f::setX(float newX) {
    this->x = newX;
    return *this;
}

float Geography::Vector2f::getY() const {
    return y;
}

Geography::Vector2f &Geography::Vector2f::setY(float newY) {
    this->y = newY;
    return *this;
}

Geography::Vector2f &Geography::Vector2f::setXY(float newX, float newY) {
    return setX(newX).setY(newY);
}

[[maybe_unused]] Geography::Vector2f &Geography::Vector2f::setXY(float newXY) {
    return setXY(newXY, newXY);
}

File: java/src/main/java/Vector2f.java
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.jetbrains.annotations.NotNull;

@Accessors(chain = true)
@Getter
@Setter
public class Vector2f {
    private float x, y;
    
    public Vector2f() { this(0); }
    
    public Vector2f(float xy) { this(xy, xy); }
    
    public Vector2f(@NotNull Vector2f vector2f) { this(vector2f.getX(), vector2f.getY()); }
    
    public Vector2f(float x, float y) { setXY(x, y); }
    
    public Vector2f setXY(float x, float y) { return setX(x).setY(y); }
    
    public Vector2f add(@NotNull Vector2f vector2f) { return add(vector2f, false); }
    
    public Vector2f add(@NotNull Vector2f vector2f, boolean replace) {
        if (replace) {
            setXY(getX() + vector2f.getX(), getY() + vector2f.getY()); return this;
        } else {
            return new Vector2f(getX() + vector2f.getX(), getY() + vector2f.getY());
        }
    }
    
    public Vector2f sub(@NotNull Vector2f vector2f) { return add(vector2f, false); }
    
    public Vector2f sub(@NotNull Vector2f vector2f, boolean replace) {
        if (replace) {
            setXY(getX() - vector2f.getX(), getY() - vector2f.getY()); return this;
        } else {
            return new Vector2f(getX() + vector2f.getX(), getY() + vector2f.getY());
        }
    }
    
    public Vector2f addX(float x) { setX(getX() + x); return this; }
    
    public Vector2f addX(@NotNull Vector2f vector2f) { setX(getX() + vector2f.getX()); return this; }
    
    public Vector2f addY(float y) { setY(getY() + y); return this; }
    
    public Vector2f addY(@NotNull Vector2f vector2f) { setY(getY() + vector2f.getY()); return this; }
    
    @Override
    public String toString() {
        return String.format("(%.2f, %.2f)", getX(), getY());
    }
}
